archive/imo/imo2014_q1.lean

```diff
@@ -0,0 +1,183 @@
+/-
+Copyright (c) 2021 Adrián Doña Mateo. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adrián Doña Mateo
+-/
+
+import algebra.big_operators
+import data.pnat.basic
+
+/-!
+# IMO 2014 Q1
+
+Let a₀ < a₁ < a₂ < ⋯ be an infinite sequence of positive integers.
+Prove that there exists a unique n ≥ 1 such that
+
+  aₙ < (a₀ + a₁ + ⋯ + aₙ) / n < aₙ₊₁.
+
+This solutions is a translation of the official solution, which may be found as the
+solution to problem A1 [here](https://www.imo-official.org/problems/IMO2014SL.pdf).
+-/
+
+open_locale big_operators
+
+noncomputable theory
+
+/- The integer sequence. -/
+variable (a : ℕ → ℤ)
+
+/- We define an auxiliary sequence `d a : ℕ+ → ℤ` by
+  `d a n = (a 0 + a 1 + ⋯ + a n) - n * (a n)`. -/
+def d (n : ℕ+) : ℤ := (∑ i : fin (n + 1), a i) - n * (a n)
+
+lemma first_ineq_iff {n : ℕ+} :
+  (a n : ℚ) < (↑∑ i : fin (n + 1), a i : ℚ) / n ↔ 0 < d a n :=
+begin
+  have : (↑(a n) : ℚ) = ↑(a n) / ↑(1 : ℤ) := by simp,
+  change _ < _ / ↑(n : ℤ) ↔ _,
+  rw [this, rat.div_lt_div_iff_mul_lt_mul], swap,
+  { norm_num }, swap,
+  { norm_cast, have : 1 ≤ ↑n := pnat.one_le n, linarith },
+  simp [d], rw mul_comm
+end
+
+lemma second_ineq_iff {n : ℕ+} :
+  (↑∑ i : fin (n + 1), a i : ℚ) / n ≤ a (n + 1) ↔ d a (n + 1) ≤ 0 :=
+begin
+  have : (↑(a (n + 1)) : ℚ) = ↑(a (n + 1)) / ↑(1 : ℤ) := by simp,
+  change _ / ↑(n : ℤ) ≤ _ ↔ _, rw this,
+  simp [rat.div_num_denom],
+  rw rat.le_def, swap,
```

Comment:
Does it work to put `rat.le_def inside the `simp`? Also, this is a non-terminal simp, so you should use `squeeze_simp` to replace it by `simp only`

Reply:
I couldn't get it to work inside the simp. I'm not sure, but perhaps this is because `rw rat.le_def` creates two extra subgoals for its assumptions?